code/datums/components/food/decomposition.dm

```diff
@@ -0,0 +1,162 @@
+
+	RegisterSignal(parent, COMSIG_MOVABLE_MOVED, .proc/handle_movement)
+	RegisterSignal(parent, list(
+		COMSIG_ITEM_PICKUP, //person picks up an item
+		COMSIG_STORAGE_ENTERED), //Object enters a storage object (boxes, etc.)
+		.proc/picked_up)
+	RegisterSignal(parent, list(
+		COMSIG_ITEM_DROPPED, //Object is dropped anywhere
+		COMSIG_STORAGE_EXITED), //Object exits a storage object (boxes, etc)
+		.proc/dropped)
+	RegisterSignal(parent, COMSIG_PARENT_EXAMINE, .proc/examine) // Self-explanitory
+
+	if(decomp_flags & GROSS && !(decomp_flags & RAW))
+		time_remaining = DECOMPOSITION_TIME_GROSS
+		examine_type = 1
+	if(decomp_flags & RAW) // Raw food overrides gross
+		time_remaining = DECOMPOSITION_TIME_RAW
+		examine_type = 2
+
+/datum/component/decomposition/UnregisterFromParent()
+	UnregisterSignal(parent, list(
+		COMSIG_ITEM_PICKUP,
+		COMSIG_STORAGE_ENTERED,
+		COMSIG_MOVABLE_MOVED,
+		COMSIG_ITEM_DROPPED,
+		COMSIG_STORAGE_EXITED,
+		COMSIG_PARENT_EXAMINE))
+
+/datum/component/decomposition/proc/handle_movement() // Assuming it is guaranteed to be food
+	var/obj/item/food/food = parent
+	var/atom/last_loc = food.loc
+
+	var/clean = FALSE // Used to check if it's on a clean surface
+	while(last_loc && !isarea(last_loc))
+		if(HAS_TRAIT(food.loc, TRAIT_PROTECT_FOOD))
+			clean = TRUE
+			break
+		else if(locate(/obj/structure/table) in food.loc)
+			clean = TRUE
+			break
+		last_loc = last_loc.loc
+	if(!handled || clean || protected)
+		// prevent decomposition
+		if(active_timers)
+			time_remaining = timeleft(timerid)
+			deltimer(timerid)
+		return
+	// do decomposition
+	timerid = addtimer(CALLBACK(src, .proc/decompose), time_remaining, TIMER_STOPPABLE | TIMER_UNIQUE) //Was told to use this instead of processing I guess
+
+/datum/component/decomposition/Destroy()
+	if(active_timers)
+		deltimer(timerid) //Just in case
+	return ..()
+
+/datum/component/decomposition/proc/dropped()
+	SIGNAL_HANDLER
+	protected = FALSE
+	handle_movement()
+
+/datum/component/decomposition/proc/picked_up()
+	SIGNAL_HANDLER
+	if(active_timers)
+		time_remaining = timeleft(timerid)
+		deltimer(timerid)
+	protected = TRUE
+	if(!handled)
+		handled = TRUE
+
+/datum/component/decomposition/proc/decompose()
+	var/obj/item/food/decomp = parent //Lets us spawn things at decomp
+	new /obj/effect/decal/cleanable/ants(decomp.loc)
+	new /obj/item/food/badrecipe/moldy(decomp.loc)
+	decomp.visible_message("<span class='notice'>[decomp] gets overtaken by mold and ants! Gross!</span>")
+	qdel(decomp)
+	return
+
+/datum/component/decomposition/proc/examine(datum/source, mob/user, list/examine_list)
+	SIGNAL_HANDLER
+	if(active_timers) // Is the timer currently applied to this?
+		switch(examine_type)
+			if(0)// All other types
+				switch(timeleft(timerid)) // Deciseconds used so there's no gaps between examine times.
+					if(3001 to 4500) // 7.5 to 5 Minutes left
+						examine_list += "[parent] looks kinda stale."
+					if(1501 to 3000) // 5 to 2.5 Minutes left
```

Comment:
I notice that these two switch statements, the ones for "this has a running timer" and "this is stale" look quite similar. Is the only difference that one uses the timer time, and the other uses the time remaining value?

Reply:
Yep

Reply:
Why not do something like this
```
var/time = 0
if(active_timers)
    time = timeleft(timerid)
else
    whatever the other condition is
switch(time)
```

Reply:
that would probably be very smart and efficient, which explains why I hadn't thought of it myself. I'll do that.

Reply:
It's just cause I've done this sort of thing a ton, experience comes with practice. Don't be a dick to yourself